Assets/Scripts/Framework/StateMachine/StateMachine.cs

```diff
@@ -0,0 +1,252 @@
+        /// The target state.
+        /// </summary>
+        public State TargetState { get; private set; }
+
+        /// <summary>
+        /// The previous state.
+        /// </summary>
+        public State PreviousState
+        {
+            get
+            {
+                if(_navigationHistory.Count < 2)
+                    return null;
+
+                if (!TryGetState(_navigationHistory[^2], out State state))
+                    return null;
+
+                return state;
+            }
+        }
+
+        /// <summary>
+        /// Checks if the given state is the current state.
+        /// </summary>
+        /// <typeparam name="T">The type of the state to check.</typeparam>
+        /// <returns>True if the given state is the current state, false otherwise.</returns>
+        public bool IsCurrentState<T>() where T : State => CurrentState.GetType() == typeof(T);
+
+        /// <summary>
+        /// Checks if the given state is the current state.
+        /// </summary>
+        /// <param name="state">The type of the state to check.</param>
+        /// <returns>True if the given state is the current state, false otherwise.</returns>
+        public bool IsCurrentState(State state)
+            => CurrentState.GetType() == state.GetType();
+
+        /// <summary>
+        /// Goes to the given state.
+        /// </summary>
+        /// <typeparam name="T">The target state.</typeparam>
+        /// <param name="addToHistory">Whether to add the state to the navigation history.</param>
+        public void GoToState<T>(bool addToHistory = true) where T : State
+            => GoToState(typeof(T), addToHistory);
+
+        /// <summary>
+        /// Goes to the given state.
+        /// </summary>
+        /// <param name="state">The target state.</param>
+        /// <param name="addToHistory">Whether to add the state to the navigation history.</param>
+        public void GoToState(State state, bool addToHistory = true)
+            => GoToState(state.GetType(), addToHistory);
+
+        /// <summary>
+        /// Goes to the given state.
+        /// </summary>
+        /// <param name="state">The target state</param>
+        /// <param name="addToHistory">Whether to add the state to the navigation history.</param>
+        public void GoToState(Type state, bool addToHistory = true)
+        {
+            if (CurrentState != null && CurrentState.GetType() == state)
+            {
+                Debug.LogError($"State of type {state} is already active");
+                return;
+            }
+
+            if (!TryGetState(state, out State targetState))
+            {
+                Debug.LogError($"State of type {state} is not registered");
+                return;
+            }
+
+            StartCoroutine(TransitionToState(targetState, addToHistory));
+        }
+
+        /// <summary>
+        /// Goes back to the previous state.
+        /// </summary>
+        public void GoBack()
```

Comment:
Might be preference but isn't GoBack a bit generic? Perhaps GoBackToPreviousState would be better, or RevertToPreviousState.

Reply:
Resolved